asteroid_filterbanks/pcen.py

```
@@ -0,0 +1,181 @@
+        n_channels: int = 1,
+        trainable: Union[bool, TrainableParameters] = False,
+        per_channel_smoothing: bool = False,
+    ):
+        super().__init__()
+
+        if trainable is True or trainable is False:
+            trainable = TrainableParameters(
+                alpha=trainable, delta=trainable, root=trainable, smooth=trainable
+            )
+
+        self.alpha = nn.Parameter(
+            torch.full((n_channels,), fill_value=alpha), requires_grad=trainable["alpha"]
+        )
+        self.delta = nn.Parameter(
+            torch.full((n_channels,), fill_value=delta), requires_grad=trainable["delta"]
+        )
+        self.root = nn.Parameter(
+            torch.full((n_channels,), fill_value=root), requires_grad=trainable["root"]
+        )
+
+        self.floor = floor
+        self.ema = ExponentialMovingAverage(
+            smooth=smooth,
+            per_channel=per_channel_smoothing,
+            n_channels=n_channels,
+            trainable=trainable["smooth"],
+        )
+
+    def forward(
+        self, tf_rep: torch.Tensor, initial_state: Optional[torch.Tensor] = None
+    ) -> Tuple[torch.tensor, torch.Tensor]:
+        """Computes the PCEN from a complex frequency representation.
+
+        Args:
+            tf_rep: A tensor containing an asteroid complex time frequency representation.
+            initial_state: A tensor containing the initial hidden state.
+                If set to None, defaults to the first element by time (dim=-1) in the tf_rep tensor.
+                Defaults to None
+
+        Shapes
+            >>> (batch, n_channels, freq, time), None -> (batch, n_channels, freq // 2 + 1, time), (batch_size, n_channel, freq // 2 + 1, 1)
+            >>> (batch, n_channels, freq, time), (batch_size, n_channel, freq // 2 + 1, 1) -> (batch, n_channels, freq // 2 + 1, time), (batch_size, n_channel, freq // 2 + 1, 1)
+
+        """
+        if not transforms.is_asteroid_complex(tf_rep):
```

Why it expects a complex input ? 
In the original paper https://arxiv.org/pdf/1607.05666.pdf, the input is Mel spectra if i am not wrong. 
I would not check anything about the input exceprt the shape.

---

Hmm, good point. 
It can be any magnitude spectrum (not specifically a mel) 

```python
>>> audio = torch.randn(10, 2, 16_000)
>>> fb = STFTFB(kernel_size=256, n_filters=256, stride=128)
>>> enc = Encoder(fb)
>>> tf_rep = enc(audio)
>>> pcen = PCEN(n_channels=2)
>>> mag_spec = transform.mag(tf_rep)
>>> energy = pcen(mag_spec)
```
or 

```python
>>> audio = torch.randn(10, 2, 16_000)
>>> mel = MelScale(...)
>>> mag_spec = mel(audio)
>>> energy = pcen(mag_spec)
```

Will simplifiy!

---

yeah checking for the input to be real could be good

---

@popcornell I'm actually a bit unsure how to check if they are real or not. Any guidance on this?

---

Maybe you could raise an error if is_asteroid_complex is True or input.dtype in [torch.complex64, torch.complex128]

---

I mean it is not strictly required to check this, i think it will fail nonetheless without the check

---

I'm not sure any type of check is needed.